clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -179,14 +179,16 @@ public double measure(MetricConfig config, long now) {
      * @param headers the Headers for the record
      * @param callback The user-supplied callback to execute when the request is complete
      * @param maxTimeToBlock The maximum time in milliseconds to block for buffer memory to be available
+     * @param retryOnNewBatch A boolean that indicates whether we return and run onNewBatch before trying to append again
      */
     public RecordAppendResult append(TopicPartition tp,
                                      long timestamp,
                                      byte[] key,
                                      byte[] value,
                                      Header[] headers,
                                      Callback callback,
-                                     long maxTimeToBlock) throws InterruptedException {
+                                     long maxTimeToBlock,
+                                     boolean retryOnNewBatch) throws InterruptedException {
```

Comment:
I think this should be something like `abortOnBatchFull` to convey the fact that we want `append` to abort if the existing batch is full.  Retrying is what the calling code does, not what `append` does.